README
NAME
    App::autotest - main package for the autotest tool

VERSION
    version 0.005

AUTHOR
    Gregor Goldbach <glauschwuffel@nomaden.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Gregor Goldbach.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
0.005     2013-04-08 20:14:48 Europe/Berlin

  [BUGS]

  Updated dist.ini: We need a minimum version 0.46 Test::Spec since we use
  the method "with" which is introduced in that version.

0.003     2012-11-13 22:08:29 Europe/Berlin

  [BUGS]

  Corrected a minor bug in a test: I guessed the order of directory entries.
  Thanks to CPAN Testers this was caught.

  The POD for autotest contained duplicate entries since I typed (manually)
  what Dist::Zilla adds (automatically).

0.002

  [ENHANCEMENTS]

  When the last run was red and the current run is green we're told that
  things just got better.

0.001

initial version


lib/App/autotest/Test/Runner.pm
package App::autotest::Test::Runner;
{
  $App::autotest::Test::Runner::VERSION = '0.005';
}

# ABSTRACT: runs tests

use strict;
use warnings;

use Moose;
use TAP::Harness;
use App::autotest::Test::Runner::Result;

has harness => (
  is      => 'rw',
  isa     => 'TAP::Harness',
  default => sub { _default_harness() }
);

has result => (
  is  => 'rw',
  isa => 'App::autotest::Test::Runner::Result'
);

sub run {
  my ( $self, @tests ) = @_;

  my $harness_result = $self->harness->runtests(@tests);
  my $result =
    App::autotest::Test::Runner::Result->new( harness_result => $harness_result );
  $self->result($result);
}

sub had_failures {
  my ($self)=@_;

  return $self->result->has_failures;
}


sub _default_harness {
    my $args = {
        verbosity => -3,
        lib       => [ 'lib', 'blib/lib', 'blib/arch' ],
    };
    return TAP::Harness->new($args);
}

1;

__END__
=pod

=head1 NAME

App::autotest::Test::Runner - runs tests

=head1 VERSION

version 0.005

=head1 INTERNAL METHODS

=head1 AUTHOR

Gregor Goldbach <glauschwuffel@nomaden.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Gregor Goldbach.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/App/autotest/Test/Runner/Result.pm
package App::autotest::Test::Runner::Result;
{
  $App::autotest::Test::Runner::Result::VERSION = '0.005';
}

# ABSTRACT: represents the result of a test run

use strict;
use warnings;

use Moose;

has harness_result => (
  is      => 'rw',
  isa     => 'TAP::Parser::Aggregator'
);

sub has_failures {
  my ($self)=@_;

  return $self->harness_result->failed > 0;
}

1;

__END__
=pod

=head1 NAME

App::autotest::Test::Runner::Result - represents the result of a test run

=head1 VERSION

version 0.005

=head1 AUTHOR

Gregor Goldbach <glauschwuffel@nomaden.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Gregor Goldbach.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/App/autotest/Test/Runner/Result/History.pm
package App::autotest::Test::Runner::Result::History;
{
  $App::autotest::Test::Runner::Result::History::VERSION = '0.005';
}

# ABSTRACT: collects test runner results

use strict;
use warnings;

use Moose;
use App::autotest::Test::Runner::Result;

has current_result => ( is => 'rw' );
has last_result => ( is => 'rw' );


sub perpetuate {
  my ( $self, $result ) = @_;

  $self->last_result( $self->current_result ) if $self->current_result;
  $self->current_result($result);
}


sub things_just_got_better {
  my ( $self ) = @_;

  # we can't claim 'better' if we have no last result
  return unless $self->last_result;

  my $was_red=$self->last_result->has_failures;
  my $is_green=not $self->current_result->has_failures;

  return $was_red && $is_green;
}

1;

__END__
=pod

=head1 NAME

App::autotest::Test::Runner::Result::History - collects test runner results

=head1 VERSION

version 0.005

=head2 perpetuate ($result)

Stores C<$result> as the new current result.
Shifts the former current result to the last result.

=head2 things_just_got_better

Things are better if the last run was red and the current run is green.

=head1 AUTHOR

Gregor Goldbach <glauschwuffel@nomaden.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Gregor Goldbach.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/App/autotest.pm
package App::autotest;
{
  $App::autotest::VERSION = '0.005';
}

# ABSTRACT: main package for the autotest tool

use strict;
use warnings;

use Moose;
use File::Find;
use File::Spec;
use Cwd;
use File::ChangeNotify;

use App::autotest::Test::Runner;
use App::autotest::Test::Runner::Result::History;

has test_directory => ( is => 'rw', isa => 'Str', default => 't' );

has watcher => (
    is      => 'rw',
    isa     => 'File::ChangeNotify::Watcher',
    default => sub {
        File::ChangeNotify->instantiate_watcher(
            directories => ['t'],
            filter      => qr/\.t$/
        );
    }
);

has after_change_or_new_hook => (
    is      => 'rw',
    isa     => 'CodeRef',
    default => sub {
        sub { 0 }
    }
);

has history => ( is => 'rw',
    default => sub { App::autotest::Test::Runner::Result::History->new } );

has test_runner => ( is => 'rw',
    default => sub { App::autotest::Test::Runner->new });

sub run {
    my ($self) = @_;

    $self->run_tests_upon_startup;
    $self->run_tests_upon_change_or_creation;
}

sub run_tests_upon_startup {
    my ($self) = @_;

    my $all_test_programs = $self->all_test_programs( $self->test_directory );

    $self->run_tests(@$all_test_programs);
}

sub run_tests_upon_change_or_creation {
    my ($self) = @_;

    while (1) {
        $self->run_tests( @{ $self->changed_and_new_files } );

        last if $self->after_change_or_new_hook->();
    }
    return 1;
}

sub changed_and_new_files {
    my ($self) = @_;

    my @files;
    for my $event ( $self->watcher->wait_for_events() ) {
        my $type = $event->type();
        my $file_changed = $type eq 'create' || $type eq 'modify';
        push @files, $event->path() if $file_changed;
    }

    return \@files;
}

{
    my @files;

    sub all_test_programs {
        my ($self) = @_;

        @files = ();    # throw away result of last call
        find( { wanted => \&_wanted, no_chdir => 1 },
            './' . $self->test_directory );

        return \@files;
    }

    sub _wanted {
        my $cwd  = getcwd();
        my $name = $File::Find::name;

        push @files, File::Spec->catfile( $cwd, $name ) if $name =~ m{\.t$};
    }

}

sub run_tests {
    my ($self, @tests)=@_;

    my $result=$self->test_runner->run(@tests);
    $self->history->perpetuate($result);

    if ($self->history->things_just_got_better) {
        $self->print("Things just got better.\n");
    }
}

sub print {
    my ($self, @rest)=@_;
    print @rest;
}

1;

__END__
=pod

=head1 NAME

App::autotest - main package for the autotest tool

=head1 VERSION

version 0.005

=head1 AUTHOR

Gregor Goldbach <glauschwuffel@nomaden.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Gregor Goldbach.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Run test programs as soon as they are modified",
  "AUTHOR" => "Gregor Goldbach <glauschwuffel\@nomaden.org>",
  "BUILD_REQUIRES" => {
    "File::Basename" => 0,
    "TAP::Parser::Aggregator" => 0,
    "Test::Differences" => 0,
    "Test::Spec" => 0,
    "constant" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "App-autotest",
  "EXE_FILES" => [
    "scripts/autotest"
  ],
  "LICENSE" => "perl",
  "NAME" => "App::autotest",
  "PREREQ_PM" => {
    "Cwd" => 0,
    "File::ChangeNotify" => 0,
    "File::Find" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "Moose" => 0,
    "TAP::Harness" => 0,
    "Test::Spec" => "0.46",
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.005",
  "test" => {
    "TESTS" => "t/*.t t/features/*.t t/integration/*.t t/unit/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



